### PR TITLE
Feat: Show Nested Table Columns

### PR DESCRIPTION
**Pull request by danwom**
### Summary of Changes

In addition to showing nested columns as strings in a modal, we now support showing nested columns as individual rows in the list of columns. Also fixes some bugs where sorting columns would mess up descriptions and nested types.

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes.
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

File: frontend/amundsen_application/static/js/features/ColumnList/ColumnType/parser.ts [hunk 1]
```diff
@@ -91,8 +97,15 @@ function parseNestedTypeHelper(
...
 
+        const nestedString = columnType.substring(
+          startIndex,
+          nextStartIndex - 1
+        );
+        const spaceIndex = nestedString.indexOf(' ');
         children.push({
           head: columnType.substring(startIndex, currentIndex + 1),
+          name: (spaceIndex !== -1) ? nestedString.substring(0, spaceIndex) : '',
+          col_type: nestedString.substring(spaceIndex + 1),
```
Comment (danwom) [inline frontend/amundsen_application/static/js/features/ColumnList/ColumnType/parser.ts:108]: Got a bug where sometimes the last nested column is excluded from the col_type string. Need to check this substring.

File: frontend/amundsen_application/static/js/ducks/tableMetadata/api/helpers.ts [hunk 2]
```diff
@@ -3,13 +3,15 @@ import * as qs from 'simple-query-string';
 import { filterFromObj } from 'ducks/utilMethods';
 
 import {
+  NestedTableColumn,
   NotificationType,
-  PeopleUser,
+  PeopleUser, TableColumn, TableColumnType,
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/ducks/tableMetadata/api/helpers.ts:8]: The commit hook didn't work for you?
You might need to nuke node_modules and install again

File: frontend/amundsen_application/static/js/ducks/tableMetadata/api/helpers.ts [hunk 3]
```diff
@@ -32,16 +34,31 @@ export function getRelatedDashboardSlug(key: string): string {
...
  */
 export function getTableDataFromResponseData(
   responseData: API.TableDataAPI
 ): TableMetadata {
-  return filterFromObj(responseData.tableData, [
+  const tableData = filterFromObj(responseData.tableData, [
     'owners',
     'tags',
   ]) as TableMetadata;
+  parseNestedColumns(responseData.tableData.columns);
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/ducks/tableMetadata/api/helpers.ts:60]: Could we change this so we don't mutate the original array?
I would change the 'forEach' into a map and save the results.

File: frontend/amundsen_application/static/js/features/ColumnList/ColumnType/parser.ts [hunk 4]
```diff
@@ -150,6 +166,55 @@ export function parseNestedType(
...
+      });
+      const nestedChildren = convertNestedTypeToColumns(child, nestedLevel + 1);
+      nestedColumns.push(...nestedChildren);
+    }
+  });
+  // Need to re-establish the sort order since this is built recursively.
+  nestedColumns.forEach((column, index) => {
+    column.sort_order = index;
+  });
+  return nestedColumns;
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/features/ColumnList/ColumnType/parser.ts:214]: I would make the same comment as above. I hardly make any use of .forEach or .push these days, unless there is a clear performance gain.

File: frontend/amundsen_application/static/js/features/ColumnList/index.tsx [hunk 5]
```diff
@@ -237,45 +243,71 @@ const ColumnList: React.FC<ColumnListProps> = ({
...
+              <div className={`nesting-arrow-spacer spacer-${nestedLevel}`} />
+              <NestingArrow />
+            </>
+          );
+        }
+        return (
+          <>
+            {arrow}
+            <div className="column-name-container">
+              <div className="column-name">{title}</div>
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/features/ColumnList/index.tsx:305]: Could we use an 'h' here instead?
And maybe a p below.

File: frontend/amundsen_application/static/js/features/ColumnList/styles.scss [hunk 6]
```diff
@@ -57,6 +59,33 @@ $description-max-width-small: 300px;
     }
   }
 
+  .nesting-arrow-spacer {
+    height: 24px;
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/features/ColumnList/styles.scss:63]: $nested-arrow-spacer ?

File: frontend/amundsen_application/static/js/ducks/tableMetadata/api/helpers.ts [hunk 7]
```diff
@@ -32,16 +34,31 @@ export function getRelatedDashboardSlug(key: string): string {
   return encodeURIComponent(key);
 }
 
+/**
+ *
+ * @param columns
+ */
+export function parseNestedColumns(columns: TableColumn[]) {
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/ducks/tableMetadata/api/helpers.ts:41]: Let's test all these helper functions thoroughly.

File: frontend/amundsen_application/static/js/features/ColumnList/ColumnType/parser.ts (hunk 4 above)
Comment (danwom) [reply frontend/amundsen_application/static/js/features/ColumnList/ColumnType/parser.ts:225]: Normally I would map as well, but this is not a 1:1 mapping. We're reaching into the nested children and flattening them as well.

File: frontend/amundsen_application/static/js/components/Table/index.tsx [hunk 8]
```diff
@@ -280,7 +283,7 @@ const Table: React.FC<TableProps> = ({
               })}
           </>
         </tr>
-        {expandRow ? (
+        {expandRow? (
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Table/index.tsx:286]:
```suggestion
        {expandRow ? (
```

File: frontend/amundsen_application/static/js/ducks/tableMetadata/api/helpers.ts [hunk 9]
```diff
@@ -32,16 +37,32 @@ export function getRelatedDashboardSlug(key: string): string {
...
 }
 
+/**
+ *
+ * @param columns
+ */
+export function parseNestedColumns(columns: TableColumn[]): TableColumn[] {
+  return columns.map((column) => {
+    const nestedType = parseNestedType(column.col_type);
+    return {
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/ducks/tableMetadata/api/helpers.ts:47]: Nit: ```suggestion
    const nestedType = parseNestedType(column.col_type);
    
    return {
```

File: frontend/amundsen_application/static/js/features/ColumnList/ColumnType/parser.spec.ts [hunk 10]
```diff
@@ -227,3 +271,128 @@ describe('parseNestedType', () => {
...
+        col_type: 'string',
+        description: '',
+        name: 'id',
+        sort_order: 7,
+        nested_level: 2,
+        is_editable: false,
+        stats: [],
+      },
+    ];
+    const result = convertNestedTypeToColumns(nestedType);
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/features/ColumnList/ColumnType/parser.spec.ts:395]:
```suggestion
    const result = Parser.convertNestedTypeToColumns(nestedType);
```

File: frontend/amundsen_application/static/js/features/ColumnList/index.tsx [hunk 11]
```diff
@@ -226,56 +235,88 @@ const ColumnList: React.FC<ColumnListProps> = ({
...
+        }
+        return (
+          <>
+            {arrow}
+            <div className="column-name-container">
+              <h3 className="column-name">{title}</h3>
+              <p className="column-desc truncated">{description}</p>
+            </div>
+          </>
+        );
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/features/ColumnList/index.tsx:318]: I would just do:
```suggestion
        return (
          <>
            {nestedLevel > 0 && (<>
              <div className={`nesting-arrow-spacer spacer-${nestedLevel}`} />
              <NestingArrow />
            </>)}
            <div className="column-name-container">
              <h3 className="column-name">{title}</h3>
              <p className="column-desc truncated">{description}</p>
            </div>
          </>
        );
```

File: frontend/amundsen_application/static/js/features/ColumnList/styles.scss [hunk 12]
```diff
@@ -57,6 +59,33 @@ $description-max-width-small: 300px;
     }
   }
 
+  .nesting-arrow-spacer {
+    height: $nested-arrow-spacer;
+    width: $nested-arrow-spacer-max;
+    display: inline-block;
+
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/features/ColumnList/styles.scss:66]: You could do something like:
```
@for $i from 0 through 4 {
  &.spacer-{$i+1} {
      width: $nested-arrow-spacer * {$i};
    }
}
```

**Review by Golodhros — COMMENTED** [review]
Looking good!

There seems to be an issue with your prettier or your pre-commit hooks, as there are some formatting issues.

**Review by Golodhros — APPROVED** [review]
LGTM

**Comment by markgrover** [comment]
This is awesome, do you have a screenshot here that you could share?

**Comment by NiravLangaliya** [comment]
I have one request, Can we have sample data as well ? because we are getting such a good features but can't get visual or understanding better without having sample data.

**Pull request merged**